src/librustc_span/lib.rs

```diff
@@ -883,6 +889,101 @@ impl ExternalSource {
 #[derive(Debug)]
 pub struct OffsetOverflowError;
 
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, RustcEncodable, RustcDecodable)]
+pub enum SourceFileChecksumKind {
+    Stable,
+    Md5,
+    Sha1,
+}
+
+impl SourceFileChecksumKind {
+    pub fn desc(&self) -> &str {
+        match *self {
+            SourceFileChecksumKind::Stable => "stable",
+            SourceFileChecksumKind::Md5 => "md5",
+            SourceFileChecksumKind::Sha1 => "sha1",
+        }
+    }
+}
+
+impl Default for SourceFileChecksumKind {
+    fn default() -> Self {
+        SourceFileChecksumKind::Stable
+    }
+}
+
+impl FromStr for SourceFileChecksumKind {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<SourceFileChecksumKind, ()> {
+        match s {
+            "stable" => Ok(SourceFileChecksumKind::Stable),
+            "md5" => Ok(SourceFileChecksumKind::Md5),
+            "sha1" => Ok(SourceFileChecksumKind::Sha1),
+            _ => Err(()),
+        }
+    }
+}
+
+impl ToJson for SourceFileChecksumKind {
+    fn to_json(&self) -> Json {
+        self.desc().to_json()
+    }
+}
+
+rustc_data_structures::impl_stable_hash_via_hash!(SourceFileChecksumKind);
+
+/// The hash of the on-disk source file used for debug info.
+#[derive(Clone, PartialEq, Eq, Debug, RustcEncodable, RustcDecodable)]
+pub struct SourceFileChecksum {
```

Comment:
This should derive `Copy` and not be treated different than other `Copy` types (methods on it taking `&self` is fine, however). This will signal the intent that this doesn't use some heap allocation.